csharp/src/Ice/Reference.cs

```diff
@@ -1623,6 +1650,9 @@ internal void Write(OutputStream ostr)
             Endpoints = Array.Empty<Endpoint>();
             Facet = facet;
             Identity = identity;
+            _invocationInterceptors = invocationInterceptors == null ?
+                Communicator.InvocationInterceptors :
+                Communicator.InvocationInterceptors.Concat(invocationInterceptors).ToImmutableArray();
```

Comment:
For me, executing the most-specific interceptors first - the proxy interceptors - seems more natural. Why are you doing it the other way around?

Reply:
I did it this way because seems more consistent with what we do for dispatch interceptors

https://github.com/zeroc-ice/ice/blob/e96b04c0f7dfb5d1a8077dbed7306944cf656af1/csharp/src/Ice/ObjectAdapter.cs#L159-L160

Reply:
I think the other way around would be more consistent / expected, i.e.:
app code -> per-proxy interceptor -> communicator interceptor -> network -> communicator interceptor -> OA interceptor -> app code.

Reply:
I don't have a strong preference, as long as it works the same for both dispatch and invocation interceptors. The current order seems a bit more convenient because the most specific can override settings set by the more generic.

Reply:
I think what I described above matches the dispatch behavior (communicator interceptors are executed before OA interceptors) but not the client-side behavior per your PR.

For me, the most specific should be closer to the application code.

As far as overriding, you can also see it the other way around:
 - a per-proxy interceptor may set some value
 - a per-communicator interceptor writes this value into a binary context, and uses some default value otherwise

The binary context API does not allow overrides (there are only Add methods, no Set/Remove), so interceptors writing the same binary context entry better cooperate with one another.

Reply:
I agree that for invocation interceptors the proxy's interceptors should be first.

Reply:
Fixed the code to iterate in the suggested order, and do not copy the communicator interceptors